babeld/src/ubus.c

```diff
@@ -0,0 +1,373 @@
+  void *neighbour;
+
+  neighbour = blobmsg_open_table(b, format_address(neigh->address));
+  blobmsg_add_string(b, "dev", neigh->ifp->name);
+  blobmsg_add_u32(b, "hello-reach", neigh->hello.reach);
+  blobmsg_add_u32(b, "uhello-reach", neigh->uhello.reach);
+  blobmsg_add_u32(b, "rxcost", neighbour_rxcost(neigh));
+  blobmsg_add_u32(b, "txcost", neigh->txcost);
+  blobmsg_add_u32(b, "rtt", format_thousands(neigh->rtt));
+  blobmsg_add_u32(b, "channel", neigh->ifp->channel);
+  blobmsg_add_u8(b, "if_up", if_up(neigh->ifp));
+  blobmsg_close_table(b, neighbour);
+}
+
+static void babeld_ubus_get_neighbours(struct ubus_context *ctx_local,
+                                       struct ubus_object *obj,
+                                       struct ubus_request_data *req,
+                                       const char *method,
+                                       struct blob_attr *msg) {
+  struct blob_buf b = {0};
+  struct neighbour *neigh;
+  struct neighbour_list_entry *cur, *tmp;
+  void *ipv4, *ipv6;
+  int ret;
+  LIST_HEAD(neighbour_ipv4_list);
+  LIST_HEAD(neighbour_ipv6_list);
+
+  blob_buf_init(&b, 0);
+
+  FOR_ALL_NEIGHBOURS(neigh) {
+    struct neighbour_list_entry *n =
+        calloc(1, sizeof(struct neighbour_list_entry));
+    n->neighbour = neigh;
+    if (v4mapped(neigh->address)) {
+      list_add(&n->list, &neighbour_ipv4_list);
+    } else {
+      list_add(&n->list, &neighbour_ipv6_list);
+    }
+  }
+
+  blob_buf_init(&b, 0);
+
+  ipv4 = blobmsg_open_table(&b, "IPv4");
+  list_for_each_entry_safe(cur, tmp, &neighbour_ipv4_list, list) {
+    babeld_add_neighbour_buf(cur->neighbour, &b);
+    list_del(&cur->list);
+    free(cur);
+  }
+  blobmsg_close_table(&b, ipv4);
+
+  ipv6 = blobmsg_open_table(&b, "IPv6");
+  list_for_each_entry_safe(cur, tmp, &neighbour_ipv6_list, list) {
+    babeld_add_neighbour_buf(cur->neighbour, &b);
+    list_del(&cur->list);
+    free(cur);
+  }
+  blobmsg_close_table(&b, ipv6);
+
+  ret = ubus_send_reply(ctx_local, req, b.head);
+  if (ret)
+    fprintf(stderr, "Failed to send reply: %s\n", ubus_strerror(ret));
+}
+
+static const struct ubus_method babeld_methods[] = {
+    UBUS_METHOD_NOARG("get_info", babeld_ubus_babeld_info),
+    UBUS_METHOD_NOARG("get_xroutes", babeld_ubus_get_xroutes),
+    UBUS_METHOD_NOARG("get_routes", babeld_ubus_get_routes),
+    UBUS_METHOD_NOARG("get_neighbours", babeld_ubus_get_neighbours),
+};
+
+static struct ubus_object_type babeld_object_type =
+    UBUS_OBJECT_TYPE("babeld", babeld_methods);
+
+static struct ubus_object babeld_object = {
+    .name = "babeld",
+    .type = &babeld_object_type,
+    .methods = babeld_methods,
+    .n_methods = ARRAY_SIZE(babeld_methods),
+};
+
+static bool ubus_init_object() {
+  int ret;
+
+  ret = ubus_add_object(ctx, &babeld_object);
+  if (ret) {
+    fprintf(stderr, "Failed to add object: %s\n", ubus_strerror(ret));
+    return false;
+  }
+
+  return true;
+}
+
+static bool babeld_ubus_init(void) {
```

Comment:
// Initializes the global ubus context, connecting to the bus to be able to receive and send messages.